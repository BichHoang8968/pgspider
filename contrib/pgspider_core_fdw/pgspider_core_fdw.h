/*-------------------------------------------------------------------------
 *
 * spd_fdw.h
 *		  Foreign-data wrapper for remote PostgreSQL servers
 *
 * Portions Copyright (c) 2012-2015, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		  contrib/spd_fdw/spd_fdw.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef SPD_FDW_H
#define SPD_FDW_H

#include "foreign/foreign.h"
#include "lib/stringinfo.h"
#include "nodes/pathnodes.h"
#include "utils/relcache.h"

#include "libpq-fe.h"

/* in spd_fdw.c */
extern int	spd_set_transmission_modes(void);
extern void spd_reset_transmission_modes(int nestlevel);

/* in connection.c */
extern PGconn *GetConnection(ForeignServer *server, UserMapping *user,
			  bool will_prep_stmt);
extern void ReleaseConnection(PGconn *conn);
extern unsigned int GetCursorNumber(PGconn *conn);
extern unsigned int GetPrepStmtNumber(PGconn *conn);
extern void pgfdw_report_error(int elevel, PGresult *res, PGconn *conn,
				   bool clear, const char *sql);

/* in option.c */
extern int ExtractConnectionOptions(List *defelems,
						 const char **keywords,
						 const char **values);

#endif							/* SPD_FDW_H */
