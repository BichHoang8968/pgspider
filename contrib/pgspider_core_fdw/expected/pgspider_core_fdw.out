DELETE FROM pg_spd_node_info;
--SELECT pg_sleep(15);
CREATE EXTENSION pgspider_core_fdw;
CREATE SERVER pgspider_svr FOREIGN DATA WRAPPER pgspider_core_fdw OPTIONS (host '127.0.0.1',port '50849');
CREATE USER mapping for public server pgspider_svr OPTIONS(user 'postgres',password 'postgres');
CREATE FOREIGN TABLE test1 (i int,__spd_url text) SERVER pgspider_svr;
CREATE EXTENSION postgres_fdw;
CREATE EXTENSION file_fdw;
CREATE EXTENSION sqlite_fdw;
CREATE EXTENSION tinybrace_fdw;
CREATE EXTENSION mysql_fdw;
/*
CREATE SERVER file_svr FOREIGN DATA WRAPPER file_fdw;
CREATE SERVER filesvr2 FOREIGN DATA WRAPPER file_fdw;
CREATE FOREIGN TABLE test1__file_svr__0 (i int) SERVER file_svr options(filename '/tmp/pgtest.csv');
SELECT * FROM test1;
CREATE FOREIGN TABLE test1__filesvr2__0 (i int) SERVER file_svr options(filename '/tmp/pgtest.csv');
SELECT * FROM test1 order by i,__spd_url;
SELECT * FROM test1 UNDER '/file_svr/' ORDER BY i,__spd_url;
SELECT * FROM test1 under '/file_svr/' where i = 1;
*/
CREATE SERVER tiny_svr FOREIGN DATA WRAPPER tinybrace_fdw OPTIONS (host '127.0.0.1',port '5100', dbname 'test.db');
CREATE USER mapping for public server tiny_svr OPTIONS(username 'user',password 'testuser');
CREATE FOREIGN TABLE test1__tiny_svr__0 (i int) SERVER tiny_svr OPTIONS(table_name 'test1');
SELECT * FROM test1__tiny_svr__0 ORDER BY i;
 i 
---
 1
(1 row)

SELECT * FROM test1 ORDER BY i,__spd_url;
 i | __spd_url  
---+------------
 1 | /tiny_svr/
(1 row)

SELECT * FROM test1 UNDER '/tiny_svr/';
 i | __spd_url  
---+------------
 1 | /tiny_svr/
(1 row)

SELECT * FROM test1 under '/tiny_svr/' where i = 1;
 i | __spd_url  
---+------------
 1 | /tiny_svr/
(1 row)

CREATE SERVER post_svr FOREIGN DATA WRAPPER postgres_fdw OPTIONS (host '127.0.0.1',port '15432');
CREATE USER mapping for public server post_svr OPTIONS(user 'postgres',password 'postgres');
CREATE FOREIGN TABLE test1__post_svr__0 (i int) SERVER post_svr OPTIONS(table_name 'test1');
SELECT * FROM test1__post_svr__0 ORDER BY i;
  i   
------
 1111
(1 row)

SELECT * FROM test1 ORDER BY i,__spd_url;
  i   | __spd_url  
------+------------
    1 | /tiny_svr/
 1111 | /post_svr/
(2 rows)

SELECT * FROM test1 UNDER '/post_svr/' ORDER BY i,__spd_url; 
  i   | __spd_url  
------+------------
 1111 | /post_svr/
(1 row)

SELECT * FROM test1 under '/post_svr/' where i = 1 ORDER BY i,__spd_url;
 i | __spd_url 
---+-----------
(0 rows)

CREATE SERVER sqlite_svr FOREIGN DATA WRAPPER sqlite_fdw OPTIONS (database '/tmp/pgtest.db');
CREATE FOREIGN TABLE test1__sqlite_svr__0 (i int) SERVER sqlite_svr OPTIONS(table 'test1');
SELECT * FROM test1 ORDER BY i,__spd_url;
   i   |  __spd_url   
-------+--------------
     1 | /sqlite_svr/
     1 | /tiny_svr/
  1111 | /post_svr/
 22222 | /sqlite_svr/
(4 rows)

SELECT * FROM test1 UNDER '/sqlite_svr/' ORDER BY i,__spd_url;
   i   |  __spd_url   
-------+--------------
     1 | /sqlite_svr/
 22222 | /sqlite_svr/
(2 rows)

SELECT * FROM test1 under '/sqlite_svr/' where i = 4 ORDER BY i,__spd_url;
 i | __spd_url 
---+-----------
(0 rows)

CREATE SERVER mysql_svr FOREIGN DATA WRAPPER mysql_fdw OPTIONS (host '127.0.0.1',port '3306');
CREATE USER mapping for public server mysql_svr OPTIONS(username 'root',password 'Mysql_1234');
CREATE FOREIGN TABLE test1__mysql_svr__0 (i int) SERVER mysql_svr OPTIONS(dbname 'test',table_name 'test1');
SELECT * FROM test1 ORDER BY i,__spd_url;
   i   |  __spd_url   
-------+--------------
     1 | /mysql_svr/
     1 | /sqlite_svr/
     1 | /tiny_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
  1111 | /post_svr/
 22222 | /sqlite_svr/
(9 rows)

SELECT * FROM test1 UNDER '/mysql_svr/' ORDER BY i,__spd_url;
  i  |  __spd_url  
-----+-------------
   1 | /mysql_svr/
 777 | /mysql_svr/
 777 | /mysql_svr/
 777 | /mysql_svr/
 777 | /mysql_svr/
(5 rows)

SELECT * FROM test1 where i = 1 ORDER BY i,__spd_url;
 i |  __spd_url   
---+--------------
 1 | /mysql_svr/
 1 | /sqlite_svr/
 1 | /tiny_svr/
(3 rows)

SELECT * FROM test1 under '/mysql_svr/' where i = 5 ORDER BY i,__spd_url;
 i | __spd_url 
---+-----------
(0 rows)

SELECT * FROM test1 ORDER BY i,__spd_url;
   i   |  __spd_url   
-------+--------------
     1 | /mysql_svr/
     1 | /sqlite_svr/
     1 | /tiny_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
  1111 | /post_svr/
 22222 | /sqlite_svr/
(9 rows)

SELECT * FROM test1 UNDER '/test2/' ORDER BY i,__spd_url;
 i | __spd_url 
---+-----------
(0 rows)

SELECT * FROM test1 order by i,__spd_url;
   i   |  __spd_url   
-------+--------------
     1 | /mysql_svr/
     1 | /sqlite_svr/
     1 | /tiny_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
  1111 | /post_svr/
 22222 | /sqlite_svr/
(9 rows)

SELECT * FROM test1 UNDER '/file_svr/' ORDER BY i,__spd_url;
 i | __spd_url 
---+-----------
(0 rows)

SELECT * FROM test1 under '/file_svr/' where i = 1 ORDER BY i,__spd_url;
 i | __spd_url 
---+-----------
(0 rows)

SELECT * FROM test1__tiny_svr__0 order by i;
 i 
---
 1
(1 row)

SELECT * FROM test1 ORDER BY i,__spd_url;
   i   |  __spd_url   
-------+--------------
     1 | /mysql_svr/
     1 | /sqlite_svr/
     1 | /tiny_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
  1111 | /post_svr/
 22222 | /sqlite_svr/
(9 rows)

SELECT * FROM test1 UNDER '/tiny_svr/';
 i | __spd_url  
---+------------
 1 | /tiny_svr/
(1 row)

SELECT * FROM test1 under '/tiny_svr/' where i = 1;
 i | __spd_url  
---+------------
 1 | /tiny_svr/
(1 row)

SELECT * FROM test1__post_svr__0 order by i;
  i   
------
 1111
(1 row)

SELECT * FROM test1 ORDER BY i,__spd_url;
   i   |  __spd_url   
-------+--------------
     1 | /mysql_svr/
     1 | /sqlite_svr/
     1 | /tiny_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
  1111 | /post_svr/
 22222 | /sqlite_svr/
(9 rows)

SELECT * FROM test1 UNDER '/post_svr/' ORDER BY i,__spd_url;
  i   | __spd_url  
------+------------
 1111 | /post_svr/
(1 row)

SELECT * FROM test1 under '/post_svr/' where i = 1 ORDER BY i,__spd_url;
 i | __spd_url 
---+-----------
(0 rows)

SELECT * FROM test1 ORDER BY i,__spd_url;
   i   |  __spd_url   
-------+--------------
     1 | /mysql_svr/
     1 | /sqlite_svr/
     1 | /tiny_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
  1111 | /post_svr/
 22222 | /sqlite_svr/
(9 rows)

SELECT * FROM test1 UNDER '/sqlite_svr/' ORDER BY i,__spd_url;
   i   |  __spd_url   
-------+--------------
     1 | /sqlite_svr/
 22222 | /sqlite_svr/
(2 rows)

SELECT * FROM test1 under '/sqlite_svr/' where i = 4 ORDER BY i,__spd_url;
 i | __spd_url 
---+-----------
(0 rows)

SELECT * FROM test1 ORDER BY i,__spd_url;
   i   |  __spd_url   
-------+--------------
     1 | /mysql_svr/
     1 | /sqlite_svr/
     1 | /tiny_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
  1111 | /post_svr/
 22222 | /sqlite_svr/
(9 rows)

SELECT * FROM test1 UNDER '/mysql_svr/' ORDER BY i,__spd_url;
  i  |  __spd_url  
-----+-------------
   1 | /mysql_svr/
 777 | /mysql_svr/
 777 | /mysql_svr/
 777 | /mysql_svr/
 777 | /mysql_svr/
(5 rows)

SELECT * FROM test1 where i = 1 ORDER BY i,__spd_url;
 i |  __spd_url   
---+--------------
 1 | /mysql_svr/
 1 | /sqlite_svr/
 1 | /tiny_svr/
(3 rows)

SELECT * FROM test1 under '/mysql_svr/' where i = 5 ORDER BY i,__spd_url;
 i | __spd_url 
---+-----------
(0 rows)

SELECT * FROM test1 UNION ALL SELECT * FROM test1 ORDER BY i,__spd_url;
   i   |  __spd_url   
-------+--------------
     1 | /mysql_svr/
     1 | /mysql_svr/
     1 | /sqlite_svr/
     1 | /sqlite_svr/
     1 | /tiny_svr/
     1 | /tiny_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
   777 | /mysql_svr/
  1111 | /post_svr/
  1111 | /post_svr/
 22222 | /sqlite_svr/
 22222 | /sqlite_svr/
(18 rows)

-- only __spd_url target list is OK
SELECT __spd_url FROM test1 ORDER BY __spd_url;
  __spd_url   
--------------
 /mysql_svr/
 /mysql_svr/
 /mysql_svr/
 /mysql_svr/
 /mysql_svr/
 /post_svr/
 /sqlite_svr/
 /sqlite_svr/
 /tiny_svr/
(9 rows)

SELECT sum(i) FROM test1;
  sum  
-------
 26444
(1 row)

SELECT avg(i) FROM test1;
       avg        
------------------
 2938.22222222222
(1 row)

SELECT avg(i),i FROM test1 group by i order by i;
  avg  |   i   
-------+-------
     1 |     1
   777 |   777
  1111 |  1111
 22222 | 22222
(4 rows)

SELECT sum(i),count(i),i FROM test1 group by i order by i;
  sum  | count |   i   
-------+-------+-------
     3 |     3 |     1
  3108 |     4 |   777
  1111 |     1 |  1111
 22222 |     1 | 22222
(4 rows)

CREATE FOREIGN TABLE t1 (i int, t text,__spd_url text) SERVER pgspider_svr;
CREATE FOREIGN TABLE t1__post_svr__0 (i int, t text) SERVER post_svr OPTIONS(table_name 't1');
SELECT * FROM t1;
 i | t | __spd_url  
---+---+------------
 1 | a | /post_svr/
 2 | b | /post_svr/
 3 | b | /post_svr/
 4 | c | /post_svr/
(4 rows)

SELECT sum(i),t FROM t1 group by t;
 sum | t 
-----+---
   1 | a
   5 | b
   4 | c
(3 rows)

SELECT sum(i),t,count(i) FROM t1 group by t;
 sum | t | count 
-----+---+-------
   1 | a |     1
   5 | b |     2
   4 | c |     1
(3 rows)

SELECT * FROM t1 WHERE i = 1;
 i | t | __spd_url  
---+---+------------
 1 | a | /post_svr/
(1 row)

SELECT sum(i),t FROM t1 group by t;
 sum | t 
-----+---
   1 | a
   5 | b
   4 | c
(3 rows)

SELECT avg(i) FROM t1;
 avg 
-----
 2.5
(1 row)

SELECT sum(i),t FROM t1 WHERE i = 1 group by t;
 sum | t 
-----+---
   1 | a
(1 row)

SELECT avg(i),sum(i) FROM t1;
 avg | sum 
-----+-----
 2.5 |  10
(1 row)

SELECT sum(i),sum(i) FROM t1;
 sum | sum 
-----+-----
  10 |  10
(1 row)

SELECT avg(i),t FROM t1 group by t;
 avg | t 
-----+---
   1 | a
 2.5 | b
   4 | c
(3 rows)

SELECT avg(i) FROM t1 group by i;
 avg 
-----
   3
   4
   2
   1
(4 rows)

SELECT * FROM (SELECT sum(i) FROM t1) A,(SELECT count(i) FROM t1) B;
 sum | count 
-----+-------
  10 |     4
(1 row)

PREPARE stmt AS SELECT * FROM t1;
-- First time is OK
EXECUTE stmt;
 i | t | __spd_url  
---+---+------------
 1 | a | /post_svr/
 2 | b | /post_svr/
 3 | b | /post_svr/
 4 | c | /post_svr/
(4 rows)

-- second time is crash :invalid memory alloc
-- EXECUTE stmt;
CREATE FOREIGN TABLE t3 (t text, t2 text, i int,__spd_url text) SERVER pgspider_svr;
CREATE FOREIGN TABLE t3__mysql_svr__0 (t text,t2 text,i int) SERVER mysql_svr OPTIONS(dbname 'test',table_name 'test3');
SELECT count(t) FROM t3;
 count 
-------
     4
(1 row)

SELECT count(t2) FROM t3;
 count 
-------
     3
(1 row)

SELECT count(i) FROM t3;
 count 
-------
     4
(1 row)

-- error stack test
CREATE SERVER mysql_svr2 FOREIGN DATA WRAPPER mysql_fdw OPTIONS (host '127.0.0.1',port '3306');
CREATE USER mapping for public server mysql_svr2 OPTIONS(username 'root',password 'wrongpass');
CREATE FOREIGN TABLE t3__mysql_svr2__0 (t text,t2 text,i int) SERVER mysql_svr2 OPTIONS(dbname 'test',table_name 'test3');
SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

SELECT count(t) FROM t3;
WARNING:  GetForeignRelSize failed
 count 
-------
     4
(1 row)

DROP FOREIGN TABLE t3;
DROP FOREIGN TABLE t3__mysql_svr__0;
DROP FOREIGN TABLE t3__mysql_svr2__0;
-- wrong result:
-- SELECT sum(i),t  FROM t1 group by t having sum(i) > 2;
--  sum | t 
-- -----+---
--    1 | a
--    5 | b
--    4 | c
-- (3 rows)
-- wrong result and warning:
-- SELECT t, __spd_url FROM t1 GROUP BY __spd_url, t;
-- WARNING:  dummy plan list failed
--   t | __spd_url 
--  ---+-----------
--  (0 rows)
CREATE FOREIGN TABLE t2 (i int, t text, a text,__spd_url text) SERVER pgspider_svr;
CREATE FOREIGN TABLE t2__post_svr__0 (i int, t text,a text) SERVER post_svr OPTIONS(table_name 't2');
SELECT i,t,a FROM t2 ORDER BY i,__spd_url;
 i | t | a 
---+---+---
 1 | a | 
 2 |   | a
   | a | a
   |   | 
(4 rows)

CREATE FOREIGN TABLE t2__post_svr__1 (i int, t text,a text) SERVER post_svr OPTIONS(table_name 't2');
CREATE FOREIGN TABLE t2__post_svr__2 (i int, t text,a text) SERVER post_svr OPTIONS(table_name 't2');
CREATE FOREIGN TABLE t2__post_svr__3 (i int, t text,a text) SERVER post_svr OPTIONS(table_name 't2');
SELECT i,t,a FROM t2 ORDER BY i,t,a,__spd_url;
 i | t | a 
---+---+---
 1 | a | 
 1 | a | 
 1 | a | 
 1 | a | 
 2 |   | a
 2 |   | a
 2 |   | a
 2 |   | a
   | a | a
   | a | a
   | a | a
   | a | a
   |   | 
   |   | 
   |   | 
   |   | 
(16 rows)

SELECT a,i, __spd_url, t FROM t2 ORDER BY i,t,a,__spd_url;
 a | i | __spd_url  | t 
---+---+------------+---
   | 1 | /post_svr/ | a
   | 1 | /post_svr/ | a
   | 1 | /post_svr/ | a
   | 1 | /post_svr/ | a
 a | 2 | /post_svr/ | 
 a | 2 | /post_svr/ | 
 a | 2 | /post_svr/ | 
 a | 2 | /post_svr/ | 
 a |   | /post_svr/ | a
 a |   | /post_svr/ | a
 a |   | /post_svr/ | a
 a |   | /post_svr/ | a
   |   | /post_svr/ | 
   |   | /post_svr/ | 
   |   | /post_svr/ | 
   |   | /post_svr/ | 
(16 rows)

SELECT __spd_url,i FROM t2 WHERE __spd_url='/post_svr/' ORDER BY i LIMIT 1;
 __spd_url  | i 
------------+---
 /post_svr/ | 1
(1 row)

-- Keep alive test
CREATE SERVER post_svr2 FOREIGN DATA WRAPPER postgres_fdw OPTIONS (host '127.0.0.1',port '49503');
CREATE USER mapping for public server post_svr2 OPTIONS(user 'postgres',password 'postgres');
CREATE FOREIGN TABLE t2__post_svr2__0 (i int, t text,a text) SERVER post_svr2 OPTIONS(table_name 't2');
INSERT INTO pg_spd_node_info VALUES(0,'post_svr','postgres_fdw','127.0.0.1');
INSERT INTO pg_spd_node_info VALUES(0,'post_svr2','postgres_fdw','127.0.0.1');
SELECT pg_sleep(2);
 pg_sleep 
----------
 
(1 row)

SELECT i,t,a FROM t2 ORDER BY i,t,a,__spd_url;;
 i | t | a 
---+---+---
 1 | a | 
 1 | a | 
 1 | a | 
 1 | a | 
 2 |   | a
 2 |   | a
 2 |   | a
 2 |   | a
   | a | a
   | a | a
   | a | a
   | a | a
   |   | 
   |   | 
   |   | 
   |   | 
(16 rows)

SET pgspider_core_fdw.throw_error_ifdead to true;
SELECT i,t,a FROM t2 ORDER BY i,t,a,__spd_url;;
ERROR:  PGSpider can not get data from child node : post_svr2
SET pgspider_core_fdw.throw_error_ifdead to false;
SELECT i,t,a FROM t2 ORDER BY i,t,a,__spd_url;;
 i | t | a 
---+---+---
 1 | a | 
 1 | a | 
 1 | a | 
 1 | a | 
 2 |   | a
 2 |   | a
 2 |   | a
 2 |   | a
   | a | a
   | a | a
   | a | a
   | a | a
   |   | 
   |   | 
   |   | 
   |   | 
(16 rows)

SET pgspider_core_fdw.print_error_nodes to true;
SELECT i,t,a FROM t2 ORDER BY i,t,a,__spd_url;;
WARNING:  Can not get data from post_svr2
 i | t | a 
---+---+---
 1 | a | 
 1 | a | 
 1 | a | 
 1 | a | 
 2 |   | a
 2 |   | a
 2 |   | a
 2 |   | a
   | a | a
   | a | a
   | a | a
   | a | a
   |   | 
   |   | 
   |   | 
   |   | 
(16 rows)

SET pgspider_core_fdw.print_error_nodes to false;
SELECT i,t,a FROM t2 ORDER BY i,t,a,__spd_url;;
 i | t | a 
---+---+---
 1 | a | 
 1 | a | 
 1 | a | 
 1 | a | 
 2 |   | a
 2 |   | a
 2 |   | a
 2 |   | a
   | a | a
   | a | a
   | a | a
   | a | a
   |   | 
   |   | 
   |   | 
   |   | 
(16 rows)

CREATE SERVER post_svr3 FOREIGN DATA WRAPPER postgres_fdw OPTIONS (host '192.168.11.12',port '15432');
CREATE USER mapping for public server post_svr3 OPTIONS(user 'postgres',password 'postgres');
CREATE FOREIGN TABLE t2__post_svr3__0 (i int, t text,a text) SERVER post_svr3 OPTIONS(table_name 't2');
INSERT INTO pg_spd_node_info VALUES(0,'post_svr3','postgres_fdw','192.168.11.12');
SELECT pg_sleep(2);
 pg_sleep 
----------
 
(1 row)

/*
SELECT i,t,a FROM t2 ORDER BY i,t,a,__spd_url;
SET pgspider_core_fdw.throw_error_ifdead to true;
SELECT i,t,a FROM t2 ORDER BY i,t,a,__spd_url;
SET pgspider_core_fdw.throw_error_ifdead to false;
SELECT i,t,a FROM t2 ORDER BY i,t,a,__spd_url;
SET pgspider_core_fdw.print_error_nodes to true;
SELECT i,t,a FROM t2 ORDER BY i,t,a,__spd_url;
SET pgspider_core_fdw.print_error_nodes to false;
SELECT i,t,a FROM t2 ORDER BY i,t,a,__spd_url;
DROP FOREIGN TABLE t2__post_svr3__0;
DELETE FROM pg_spd_node_info WHERE servername = 't2__post_svr3__0';
SELECT pg_sleep(2);
SELECT i,t,a FROM t2 ORDER BY i,t,a,__spd_url;
*/
DROP FOREIGN TABLE test1;
DROP FOREIGN TABLE t1;
DROP FOREIGN TABLE t2;
DROP SERVER pgspider_svr;
DROP EXTENSION pgspider_core_fdw CASCADE;
